File: solution/data_structure/2800/Main.java
// Authored by : kyj1991719
// Co-authored by : -
// Link : http://boj.kr/cb80d90792224a2395fc61cfe30f53b6

import java.util.*;
import java.lang.*;
import java.io.*;

public class Main {
    private static List<Bracket> bracketList = new ArrayList<>();
    private static List<String> eqList = new ArrayList<>();
    private static String input;

    public static void main(String[] args) {
        FastReader rd = new FastReader();

        input = rd.nextLine();
        Stack<Integer> startStack = new Stack<>();
        for(int i = 0; i < input.length(); i++) {
            if(input.charAt(i) == '(') startStack.add(i);
            if(input.charAt(i) == ')') bracketList.add(new Bracket(startStack.pop(), i));
        }

        for(int i = 0; i < bracketList.size(); i++) {
            boolean[] removedArr = new boolean[input.length()];
            removedArr[bracketList.get(i).start] = true;
            removedArr[bracketList.get(i).end] = true;
            solve(i, removedArr);
        }

        List<String> distinctList = new ArrayList<>(new HashSet<>(eqList));
        distinctList.sort((o1, o2) -> o1.compareTo(o2));
        StringBuilder sb = new StringBuilder();
        for(String tmp : distinctList) {
            sb.append(tmp);
            sb.append("\n");
        }
        System.out.print(sb.toString());
    }

    private static void solve(int idx, boolean[] removedArr) {
        if(idx == bracketList.size()) {
            addEq(removedArr);
            return;
        }

        for(int i = idx+1; i < bracketList.size(); i++) {
            Bracket bracket = bracketList.get(i);

            removedArr[bracket.start] = true;
            removedArr[bracket.end] = true;
            solve(i, removedArr);
            removedArr[bracket.start] = false;
            removedArr[bracket.end] = false;
        }

        solve(idx+1, removedArr);
    }

    private static void addEq(boolean[] removedArr) {
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < input.length(); i++) {
            if(removedArr[i]) continue;
            sb.append(input.charAt(i));
        }
        eqList.add(sb.toString());
    }


    static class Bracket {
        int start, end;
        public Bracket(int start, int end) {
            this.start = start;
            this.end = end;
        }
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while(st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() { return Integer.parseInt(next()); }
        long nextLong() { return Long.parseLong(next()); }
        double nextDouble() { return Double.parseDouble(next()); }
        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

File: solution/data_structure/2800/main.py
# Authored by : tony9402
# Co-authored by : -
# Link : http://boj.kr/9908832547364dfd87342eb0a2a873ce
import sys

def input():
    return sys.stdin.readline().rstrip()

s = input()
N = len(s)

index = [ -1 for _ in range(N) ]
stack = []
current_index = 0

# 올바른 괄호만 주어진다. 각 괄호 쌍에 차례대로 index를 붙여주고 기록하자.
for idx, ch in enumerate(s):
    if ch == '(':
        stack.append(current_index)
        index[idx] = current_index
        current_index += 1
    elif ch == ')':
        index[idx] = stack.pop()

# 각 괄호에 번호를 기록해놨다. 이제 재귀를 돌아 괄호쌍을 뽑자.
answer = [] # 가능한 모든 것을 저장 후 정렬 후 출력하기 위해 사용

# current_index ==> bracket count
choose = [ 0 for _ in range(current_index) ]

# cnt -> bracket index 번호
def func(cnt): 
    if cnt == current_index:
        erase_bracket_count = sum(choose)
        if erase_bracket_count == 0:
            return

        string = ""
        for idx, ch in enumerate(s):
            # index[idx] == -1 인 경우 (괄호가 아니므로 추가)
            # 만약 -1이면 뒤에 조건문이 실행 안됨
            if index[idx] == -1 or choose[index[idx]] == 0:
                string += ch

        answer.append(string)
        return

    choose[cnt] = 1 # 해당 괄호쌍을 지운 경우
    func(cnt + 1)
    choose[cnt] = 0 # 해당 괄호쌍을 지우지 않은 경우
    func(cnt + 1)

# Run 
func(0)
# 정답 중에 중복이 있을 수 있기 때문에 중복 제거
answer = sorted(set(answer))
print('\n'.join(answer))
